firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Protocols/datagrams.txt

```diff
@@ -0,0 +1,456 @@
+      THEN("The write method reports ok status")
+      THEN("After the write method is called, The value returned by the seq accessor method remains unchanged")
+      THEN("the value returned by the length accessor method remains unchanged")
+      THEN("The payload returned by the payload accessor method remains unchanged")
+      THEN("The seq field of the body's header matches the value returned by the seq accessor method")
+      THEN("The length field of the body's header matches the value returned by the length accessor method")
+      THEN("The body's payload section is empty")
+      THEN("The output buffer is as expected '0x00 0x00' ")
+
+  GIVEN("A Datagram constructed with a non-empty payload buffer and sequence equal to 0")
+  // assertions carried out in GIVEN:
+  // The seq accessor method returns 0
+  // The value returned by the length accessor method is 5
+  // The payload buffer returned by the payload accessor method is '0xb7 0xe1 0x8d 0xaa 0x4d'
+
+    WHEN("The sequence, length and paylaod are written to the output buffer")
+      THEN("The write method reports ok status")
+      THEN("After the write method is called, The value returned by the seq accessor method remains unchanged")
+      THEN("the value returned by the length accessor method remains unchanged")
+      THEN("The payload returned by the paylaod accessor method remains unchanged")
+      THEN("The seq field of the body's header matches the value returned by the seq accessor method")
+      THEN("The length field of the body's header matches the value returned by the length accessor method")
+      THEN("The body's payload section correctly stores the paylaod as '0xb7 0xe1 0x8d 0xaa 0x4d'")
+      THEN("The output buffer is as expected '0x00 0x05 0xb7 0xe1 0x8d 0xaa 0x4d' ")
+
+    WHEN("The paylaod given to the constructor is altered ('0xb7 0xe1 0x8d 0xaa 0x4d 0x02') before write method is called")
+      THEN("Before the write method is called, The value returned by the length accessor method is inconsistent with the size of the altered paylaod")
+      THEN("The write method reports ok status")
+      THEN("After the write method is called, The value returned by the seq accessor method remains unchanged")
+      THEN("the length accessor method returns a value equal to the size of the altered payload")
+      THEN("The paylaod returned by the paylaod accessor method is same as the altered paylaod buffer")
+      THEN("The seq field of the body's header matches the value returned by the seq accessor method")
+      THEN("The length field of the body's header matches the value returned by the length accessor method")
+      THEN("The body's payload section correctly stores the paylaod as '0xb7 0xe1 0x8d 0xaa 0x4d 0x02'")
+      THEN("The output buffer is as expected '0x00 0x06 0x01 0x02 0x03 0x04 0x05 0x02' ")
+
+  GIVEN("A Datagram constructed with a non-empty payload buffer of different lengths and sequence equal to 0")
```

Comment:
Since the buffer can only have one length at a time, I'd suggest wording this as "with a non-empty payload buffer of any length from 0 to 252"; this also is closer to the language which would be used in a mathematical proof, and here the test code just happens to prove the claim by enumerating all the cases covered by the GIVEN.